gapic-common/lib/gapic/rest/server_stream.rb

```diff
@@ -0,0 +1,98 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  module Rest
+    class ServerStream
+      include Enumerable
+
+      # @return Enumerable<String>
+      attr_reader :bodies
+
+      # @return Enumerable<String>
+      attr_reader :enumberable
+
+      # @param fiber Enumerable<String>
+      def initialize enumerable
+        @enumerable = enumerable
+        @_level = 0
+        @_obj = ""
+        @ready_objs = []
+      end
+
+      def next_json!(chunk)
```

Comment:
If we can assume that a server stream is always a toplevel JSON array, then the only "partial" parsing that we care about is that top level array. (For example, any "unfinished" objects below that are just not ready, and we can ignore them until they are ready.) So, to avoid trying to reimplement an entire JSON parser, I think we should limit our custom parsing to that top level, and then call the real Ruby JSON parser to parse objects below it.

Reply:
Hm, are you saying something like this:

```
function next_json(chunk)
  current_json += chunk
  try res = RubyJSONParser.parse(current_json)
  catch ParsingError: return # JSON not finished yet
  return res
```

?

Reply:
Not quite, because that will never succeed until the last chunk arrives (since the top level array of stream elements will not be closed until then). So you do need to do some custom parsing at the top level, but is there a way you can delegate to the [standard Ruby JSON parser](https://ruby-doc.org/stdlib-3.1.2/libdoc/json/rdoc/index.html) for everything else? Because your current JSON parser is incomplete (e.g. it does not handle escaped characters properly) and implementing JSON parsing in a performant way is actually trickier than it sounds.

Reply:
Done. Switched to standard JSON parser. Thanks for the feedback.